docs: document custom step usage (#2198)
slackapi/bolt-js: docs/content/basic/custom-steps.md

---
title: Listening and responding to custom steps
lang: en
slug: /concepts/custom-steps
---

> Your app can use the `function()` method to listen to incoming [custom step requests](https://api.slack.com/automation/functions/custom-bolt). Custom steps are used in Workflow Builder to build workflows. The method requires a step `callback_id` of type string. This `callback_id` must also be defined in your [Function](https://api.slack.com/concepts/manifests#functions) definition. Custom steps must be finalized using the `complete()` or `fail()` listener arguments to notify Slack that your app has processed the request.

Reviewer: ```suggestion
Your app can use the `function()` listener to listen to incoming [custom step requests](https://api.slack.com/automation/functions/custom-bolt). Custom steps are used in Workflow Builder to build workflows.

Adding a custom step starts with a definition in [the app manifest](https://api.slack.com/concepts/manifests#functions) and a `function()` listener listening for the `callback_id` of that function. Upon function execution the listener is called and can run whatever code, but must finish with either the `complete()` or `fail()` listener argument to end the step:
```

Hoping to set a bit more context for custom steps in just the first paragraph! Also wanting to frame the following one as more instruction instead of reference, but feel free to change it or ignore 🙏

Reviewer: Also not certain about the "method" vs "listener" naming!

Author: The rest of the documentation uses the `method` terminology instead of the `listener` 

I think its better to be consistent and stick to `method`

We can send a follow up PR to update all the terms 👍

> * `complete()` requires one argument: an `outputs` object. It completes your custom step **successfully** and provides an object containing the outputs of your custom step as per its definition.
> * `fail()` requires **one** argument: `error` of type string. It completes your custom step **unsuccessfully** and provides a message containing information regarding why your custom step failed.

Reviewer: ```suggestion
* `complete()` requires one argument: an `outputs` object. It ends your custom step **successfully** and provides an object containing the outputs of your custom step as per its definition.
* `fail()` requires **one** argument: `error` of type string. It ends your custom step **unsuccessfully** and provides a message containing information regarding why your custom step failed.
```

Trying to avoid confusion around the "complete" wording and argument 😅

You can reference your custom step's inputs using the `inputs` listener argument.

```js
// This sample custom step formats an input and outputs it

> app.function('sample_custom_step', async ({ ack, inputs, complete, fail, logger }) => {
>   try {
>     await ack();

Reviewer: ```suggestion
app.function('sample_custom_step', async ({ inputs, complete, fail, logger }) => {
  try {
```

Removing the `ack` here since this isn't provided for function executions

Author: I prefer to keep it in the docs, there are a number of issues opened related to calling ack before executing custom logic

Keeping this in hopes that it will help developers avoid a bug 🙏

We can remove it in a follow up PR as well 🤔

Reviewer: I believe the `function_executed` events are actually different from other events and can't use `ack` at all!

Within JavaScript code this is `undefined` within the listener so this call will error 😬  It's not so clear from this picture but the hover hint is hinting at this in a TypeScript project:

![undefined](https://github.com/user-attachments/assets/a05d9589-c90c-404d-8878-9a7dbd5d536a)

    const { message } = inputs;

    await complete({
        outputs: { 
            message: `:wave: You submitted the following message: \n\n>${message}` 
        }
    });
  } catch (error) {
    logger.error(error);
    await fail({ error: `Failed to handle a function request: ${error}` });
  }
});
```

> <details>

Author: I unhide this and added a header here is a preview, Let me know what you think

<img width="1365" alt="Screenshot 2024-08-14 at 12 01 41 PM" src="https://github.com/user-attachments/assets/b9c19a79-e6d5-4d49-bffb-2cf6dec4f96e">

> <details>

Reviewer 2: Love it!

  <summary>
  Listening to custom step actions
  </summary>

>   Your app can listen to user actions, like button clicks, created from `custom steps` using the `action` method.

Reviewer 2: Suggested tweak to the introduction here:
```suggestion
  Your app's custom steps may create interactivity points for users, for example:

- Post a message with a button
- [Open a modal view](creating-modals)

If such interaction points originate from a custom step execution, the events sent to your app representing the end-user interaction with these points are considered to be _function-scoped interactivity events_. These interactivity events can be handled by your app using the same concepts we covered earlier, such as [Listening to actions](action-listening) or [Listening to views](view-submissions).

The one addition that function-scoped interactivity events have is they will contain data related to the `function_executed` event they were spawned from, such as custom step input data and access to `complete()` and `fail()` handlers to manage step control flow.
```

Author: This is good 🙇  provides more information without duplicating documentation from api.slack.com

I had to adapt this slightly to because open modal views are not yet supported, but hopefully soon

>   Actions can be filtered on an `action_id` of type string or RegExp object. `action_id`s act as unique identifiers for interactive components on the Slack platform.

Reviewer 2: I'd remove this, as in my above suggestion, we can just link to the action-listening docs instead.

>   Your app can skip calling `complete()` or `fail()` in the `function()` listener if the custom step creates an `action` that waits for user interaction. However, in the `action()` method, your app must invoke `complete()` or `fail()` to notify Slack that the custom step has been processed.

Reviewer 2: ```suggestion
  Your app can skip calling `complete()` or `fail()` in the `function()` listener if the custom step creates an interaction point that requires user interaction before the step can complete. However, in the relevant interactivity handler method, your app must invoke `complete()` or `fail()` to notify Slack that the custom step has been processed.
```

>   You’ll notice in all `action()` examples, `ack()` is used. It is required to call the `ack()` function within an action listener to acknowledge that the request was received from Slack. This is discussed in the [acknowledging requests section](/concepts/acknowledge).

Reviewer 2: ```suggestion
  You’ll notice in all interactivity handler examples, `ack()` is used. It is required to call the `ack()` function within an interactivity listener to acknowledge that the request was received from Slack. This is discussed in the [acknowledging requests section](/concepts/acknowledge).
```

```js
/** This sample custom step posts a message with a button */

> app.function('sample_function', async ({ ack, client, inputs, fail, logger }) => {
>   try {
>     await ack();

Reviewer: ```suggestion
app.function('sample_function', async ({ client, inputs, fail, logger }) => {
  try {
```

* `complete()` requires one argument: an `outputs` object. It ends your custom step **successfully** and provides an object containing the outputs of your custom step as per its definition.
* `fail()` requires **one** argument: `error` of type string. It ends your custom step **unsuccessfully** and provides a message containing information regarding why your custom step failed.

You can reference your custom step's inputs using the `inputs` listener argument.

```js
// This sample custom step formats an input and outputs it
app.function('sample_custom_step', async ({ ack, inputs, complete, fail, logger }) => {
  try {
    await ack();
    const { message } = inputs;

    await complete({
        outputs: { 
            message: `:wave: You submitted the following message: \n\n>${message}` 
        }
    });
  } catch (error) {
    logger.error(error);
    await fail({ error: `Failed to handle a function request: ${error}` });
  }
});
```

<details>
<summary>

> definition

Reviewer: ```suggestion
Example app manifest definition
```

Super helpful as is! Just a suggestion for easier understanding 🙇
